[BEAM-77] Move InProcessRunner to its own module

kennknowles opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

I believe since the `DirectPipelineRunner` is still in place, this safe to do immediately. Then we can begin to run all tests on this runner in the same way we do for all runners, gaining coverage of both. Finally, we can simply delete `DirectPipelineRunner` which will expose any tests that require the `NeedsRunner` category.

kennknowles commented:
R: @tgroh 
R: @davorbonaci 

Perhaps just high level comments on the approach. Specifically, do you agree on this order of things:

1. Do this move now.
2. Leverage `TestPipeline` to immediately increase test coverage to include this runner.
3. Annotate tests appropriately with `NeedsRunner`
4. Delete `DirectPipelineRunner` from the SDK.

Expect some hiccups along the way. I'm sure that we will have to learn during set up of item 2, and that item 4 will feed back into item 3 since we will miss some.

davorbonaci commented:
Yes, I think I agree with a few minor notes, such as separating into smaller PRs.

We may be able to avoid conflating addition of new code (e.g., `NeedsRunner`) with the move itself. Could be easier to review and reason about it that way. (This is particularly because I'm not sure all tests will actually work on the new runner right now -- they may soon.)

We should coordinate with pending PRs from @tgroh. I'm fine whichever order you decide to proceed in.

Note that the official name of the runner will be `DirectRunner`, without `Pipeline` or `InProcess`. Naming-wise, I'd probably go with:
* Directory: `/runners/direct-java`
* Artifact: `direct-runner`
* Java package: `org.apache.beam.runners.direct.*`

kennknowles commented:
Done:

 - Removed the not-yet-used `NeedsRunner` commit from this PR.
 - Since that required a rebase, I also moved the commit on top of `apache/master`.
 - Named directory, artifact, and package as you describe.

This commit is now strictly a move with a `pom.xml`. I don't think it is helpful to split this into two, since the dependency analysis requires the code to be in place to guide the `pom.xml`.

From here out, the changes will need tighter coordination to bring up the new runner with its appropriate name and running all the integration tests appropriately.

kennknowles commented:
Rebased, fixed conflicts, Jenkins is happy. Please take a look!

kennknowles commented:
I believe Travis-CI failures are memory/performance issues in `InProcessPipelineRunner` not related to this PR, since there are no code changes here.

kennknowles commented:
Rebased to kick the tests.

davorbonaci commented on runners/direct-java/pom.xml:
```diff
@@ -0,0 +1,430 @@
+  <packaging>jar</packaging>
+
+  <properties>
+    <beam.version>0.1.0-incubating-SNAPSHOT</beam.version>
```
remove. use ${project.version} instead.

kennknowles replied:
Done.

davorbonaci commented on runners/direct-java/pom.xml:
```diff
@@ -0,0 +1,430 @@
+  <properties>
+    <beam.version>0.1.0-incubating-SNAPSHOT</beam.version>
+    <timestamp>${maven.build.timestamp}</timestamp>
+    <maven.build.timestamp.format>yyyy-MM-dd HH:mm</maven.build.timestamp.format>
```
can we remove these? Aren't they inherited from the parent?

kennknowles replied:
Done.

davorbonaci commented on runners/direct-java/pom.xml:
```diff
@@ -0,0 +1,430 @@
+              <goal>shade</goal>
+            </goals>
+            <configuration>
+              <shadeTestJar>true</shadeTestJar>
```
remove much of these?

kennknowles replied:
Done.

davorbonaci commented on runners/direct-java/pom.xml:
```diff
@@ -0,0 +1,430 @@
+      <version>${beam.version}</version>
+    </dependency>
+
+    <dependency>
```
why do we need all these?

kennknowles replied:
Let's trim `google-http-client` later and see why dependency analysis was OK with it. It is `compile` scope so it is definitely handled by the analysis.

davorbonaci commented on runners/direct-java/pom.xml:
```diff
@@ -0,0 +1,430 @@
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
```
probably not needed

kennknowles replied:
Done. It seems the same flag (`ignoreNonCompile`) that makes it not fail on `provided` and `runtime` dependencies - basically mandatory to set this - encompasses `test` as well. We can keep digging but maybe not for this PR.

davorbonaci commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactory.java:
```diff
@@ -15,8 +15,8 @@
-package org.apache.beam.sdk.runners.inprocess;
-import org.apache.beam.sdk.runners.inprocess.InProcessPipelineRunner.CommittedBundle;
+package org.apache.beam.runners.direct;
+import org.apache.beam.runners.direct.InProcessPipelineRunner.CommittedBundle;
```
empty line?

kennknowles replied:
Done.

davorbonaci commented on runners/direct-java/src/test/java/org/apache/beam/runners/direct/ParDoMultiEvaluatorFactoryTest.java:
```diff
@@ -119,7 +119,7 @@ public void processElement(ProcessContext c) {
     when(evaluationContext.createCounterSet()).thenReturn(counters);
 
-    org.apache.beam.sdk.runners.inprocess.TransformEvaluator<String> evaluator =
+    org.apache.beam.runners.direct.TransformEvaluator<String> evaluator =
```
why this is not an import? (ok to ignore)

kennknowles replied:
No idea. No need to import even; same package. Done.

davorbonaci commented on runners/pom.xml:
```diff
@@ -37,6 +37,7 @@
   <modules>
     <module>flink</module>
```
move up; direct runner should be first

kennknowles replied:
Done.

davorbonaci commented:
Looks mostly good. Advance LGTM.

kennknowles commented:
Based on advance LGTM I will merge if/when tests are green. Cutting deps and maven properties may have busted something so we'll see.

kennknowles commented:
Failure were only with Java 7 JDKs. The PermGen no longer exists in Java 8, so it isn't getting blown. I've adjusted the JVM parameters to settings that will hopefully alleviate the problem. I don't know why this problem does not occur when the runner is part of the SDK.

davorbonaci commented:
@tgroh, is this something you've seen before?
